In [69]:
import pandas as pd

In [70]:
df = pd.read_csv("/content/train.csv")

In [71]:
df.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  New_Price  Price  
0  126.2 bhp    5.0        NaN  12.50  
1   88.7 bhp    5.0  8.61 Lakh   4.50  
2  88.76 bhp    7.0        NaN   6.00  
3  140.8 bhp    5.0        NaN  17.74  
4   63.1 bhp    5.0        NaN   3.50

In [72]:
print(df.dtypes)

Unnamed: 0             int64
Name                  object
Location              object
Year                   int64
Kilometers_Driven      int64
Fuel_Type             object
Transmission          object
Owner_Type            object
Mileage               object
Engine                object
Power                 object
Seats                float64
New_Price             object
Price                float64
dtype: object


In [73]:
nan_count = df['New_Price'].isna().sum()
print("Total number of rows in column named price which has No price :", nan_count)

Total number of rows in column named price which has No price : 5032


In [74]:
num_rows = df.shape[0]
print("Total number of rows in the DataFrame:", num_rows)

Total number of rows in the DataFrame: 5847


In [75]:
nan_counts_by_column = df.isna().sum()
print(nan_counts_by_column)

Unnamed: 0              0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 36
Power                  36
Seats                  38
New_Price            5032
Price                   0
dtype: int64


In [76]:
df = df.drop('New_Price', axis=1)

**I have dropped new prices as this column has 5032 NAN values so it is not suggestible to replace it with mean/mode/median**

In [77]:
file_path = '/content/Clean_dataset.csv'

# Save the DataFrame to the specified file path
df.to_csv(file_path, index=False)

In [78]:
df.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  Price  
0  126.2 bhp    5.0  12.50  
1   88.7 bhp    5.0   4.50  
2  88.76 bhp    7.0   6.00  
3  140.8 bhp    5.0  17.74  
4   63.1 bhp    5.0   3.50

In [79]:
def clean_and_extract_numeric(x):
    return x.str.extract('(\d+\.?\d*)', expand=False)

# Apply the function to the specified columns
df['Mileage'] = clean_and_extract_numeric(df['Mileage'])
df['Engine'] = clean_and_extract_numeric(df['Engine'])
df['Power'] = clean_and_extract_numeric(df['Power'])

In [80]:
df.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type Mileage Engine  Power  \
0              41000    Diesel       Manual      First   19.67   1582  126.2   
1              46000    Petrol       Manual      First      13   1199   88.7   
2              87000    Diesel       Manual      First   20.77   1248  88.76   
3              40670    Diesel    Automatic     Second    15.2   1968  140.8   
4              86999    Diesel       Manual      First   23.08   1461   63.1   

   Seats  Price  
0    5.0  12.50  
1    5.0   4.50  
2    7.0   6.00  
3    5.0  17.74  
4    5.0   3.50

In [81]:
print(df.dtypes)

Unnamed: 0             int64
Name                  object
Location              object
Year                   int64
Kilometers_Driven      int64
Fuel_Type             object
Transmission          object
Owner_Type            object
Mileage               object
Engine                object
Power                 object
Seats                float64
Price                float64
dtype: object


In [82]:
#Converting object type to float type
df['Mileage'] = df['Mileage'].astype(float)
df['Engine'] = df['Engine'].astype(float)
df['Power'] = df['Power'].astype(float)
df['Seats']= df['Seats'].astype(float)

In [83]:
df.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type  Mileage  Engine  \
0              41000    Diesel       Manual      First    19.67  1582.0   
1              46000    Petrol       Manual      First    13.00  1199.0   
2              87000    Diesel       Manual      First    20.77  1248.0   
3              40670    Diesel    Automatic     Second    15.20  1968.0   
4              86999    Diesel       Manual      First    23.08  1461.0   

    Power  Seats  Price  
0  126.20    5.0  12.50  
1   88.70    5.0   4.50  
2   88.76    7.0   6.00  
3  140.80    5.0  17.74  
4   63.10    5.0   3.50

In [84]:
mean_value_Mileage = df['Mileage'].mean()
mean_value_Engine = df['Engine'].mean()
mean_value_Seats = df['Seats'].mean()
mean_value_Power = df['Power'].mean()

# Replace null values with the mean
df['Mileage'].fillna(mean_value_Mileage, inplace=True)
df['Engine'].fillna(mean_value_Engine, inplace=True)
df['Seats'].fillna(mean_value_Seats, inplace=True)
df['Power'].fillna(mean_value_Power, inplace=True)

**Replacing the nan values with mean is the most suggestible way to replace the values**

In [85]:
df['Mileage'] = df['Mileage'].round(2)

In [86]:
df['Engine'] = df['Engine'].astype(int)

In [87]:
df['Seats'] = df['Seats'].astype(int)

In [88]:
df['Power'] = df['Power'].round(2)

In [89]:
print(df.iloc[709])

Unnamed: 0                            733
Name                 Maruti Swift 1.3 VXi
Location                          Chennai
Year                                 2006
Kilometers_Driven                   97800
Fuel_Type                          Petrol
Transmission                       Manual
Owner_Type                          Third
Mileage                              16.1
Engine                               1631
Power                               113.8
Seats                                   5
Price                                1.75
Name: 709, dtype: object


In [90]:
# Perform one-hot encoding
df1 = pd.get_dummies(df, columns=['Fuel_Type', 'Transmission'])

# Display the DataFrame with one-hot encoded columns
print(df1)

      Unnamed: 0                              Name    Location  Year  \
0              1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1              2                      Honda Jazz V     Chennai  2011   
2              3                 Maruti Ertiga VDI     Chennai  2012   
3              4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4              6            Nissan Micra Diesel XV      Jaipur  2013   
...          ...                               ...         ...   ...   
5842        6014                  Maruti Swift VDI       Delhi  2014   
5843        6015          Hyundai Xcent 1.1 CRDi S      Jaipur  2015   
5844        6016             Mahindra Xylo D4 BSIV      Jaipur  2012   
5845        6017                Maruti Wagon R VXI     Kolkata  2013   
5846        6018             Chevrolet Beat Diesel   Hyderabad  2011   

      Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0                 41000      First    19.67    1582  126.2

In [91]:
df1.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0              41000      First    19.67    1582  126.20      5  12.50   
1              46000      First    13.00    1199   88.70      5   4.50   
2              87000      First    20.77    1248   88.76      7   6.00   
3              40670     Second    15.20    1968  140.80      5  17.74   
4              86999      First    23.08    1461   63.10      5   3.50   

   Fuel_Type_Diesel  Fuel_Type_Electric  Fuel_Type_Petrol  \
0                 1                   0                 0   
1                 0                   0                 1   
2                 1                   0                 0   
3                 1                   0                 0   
4                 1                   0                 0   

   Transmission_Automatic  Transmission_Manual  
0                       0                    1  
1                       0                    1  
2                       0                    1  
3                       1                    0  
4                       0                    1

In [92]:
df1.dtypes

Unnamed: 0                  int64
Name                       object
Location                   object
Year                        int64
Kilometers_Driven           int64
Owner_Type                 object
Mileage                   float64
Engine                      int64
Power                     float64
Seats                       int64
Price                     float64
Fuel_Type_Diesel            uint8
Fuel_Type_Electric          uint8
Fuel_Type_Petrol            uint8
Transmission_Automatic      uint8
Transmission_Manual         uint8
dtype: object

In [93]:
import datetime as dt
current_year = dt.datetime.now().year
df['Age'] = current_year - df['Year']

In [94]:
df.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type  Mileage  Engine  \
0              41000    Diesel       Manual      First    19.67    1582   
1              46000    Petrol       Manual      First    13.00    1199   
2              87000    Diesel       Manual      First    20.77    1248   
3              40670    Diesel    Automatic     Second    15.20    1968   
4              86999    Diesel       Manual      First    23.08    1461   

    Power  Seats  Price  Age  
0  126.20      5  12.50    8  
1   88.70      5   4.50   12  
2   88.76      7   6.00   11  
3  140.80      5  17.74   10  
4   63.10      5   3.50   10

In [95]:
file_path = '/content/Final_dataset.csv'

# Save the DataFrame to the specified file path
df.to_csv(file_path, index=False)